sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala

```diff
@@ -0,0 +1,237 @@
+          XXXX XXXX XXXX 5432
+  """,
+  since = "3.4.0",
+  group = "string_funcs"
+)
+case class MaskCcn(left: Expression, right: Expression)
+  extends MaskDigitSequence(left, right, "mask_ccn", false) {
+  def this(left: Expression) = this(left, Literal(MaskCcn.DEFAULT_FORMAT))
+  override protected def withNewChildrenInternal(
+      newInput: Expression, newFormat: Expression): MaskCcn =
+    copy(left = newInput, right = newFormat)
+}
+
+/**
+ * A function that converts a string representing a credit card number in a form updated to mask the
+ * digits in the interest of redacting the information. Returns an error if the format string is
+ * invalid or NULL if the input string does not match the format string.
+ */
+@ExpressionDescription(
+  usage = """
+    _FUNC_(input) - Convert string 'input' representing a credit card number to an updated version
+    applying a transformation to the characters. This can be useful for creating copies of tables
+    with sensitive information removed, but retaining the same schema. Returns an error if the
+    format string is invalid or NULL if the input string does not match the format string.
+    The format can consist of the following characters, case insensitive:
+      - Each 'X' represents a digit which will be converted to 'X' in the result.
+      - Each digit '0'-'9' represents a digit which will be left unchanged in the result.
+      - Each '-' character should match exactly in the input string.
+      - Each whitespace character is ignored.
+    No other format characters are allowed. Any whitespace in the input string is left unchanged.
+    The default is: XXXX-XXXX-XXXX-XXXX.
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(ccn) FROM VALUES ("1234-5678-9876-5432") AS tab(ccn);
+        XXXX-XXXX-XXXX-XXXX
+      > SELECT _FUNC_("  1234 5678 9876 5432", "XXXX XXXX XXXX 1234");
+          XXXX XXXX XXXX 5432
+      > SELECT _FUNC_("1234567898765432");
+        NULL
+      > SELECT _FUNC_("1234567898765432", "XXXX-XXXX-XXXX-1234");
+        NULL
+  """,
+  since = "3.4.0",
+  group = "string_funcs"
+)
+case class TryMaskCcn(left: Expression, right: Expression)
+  extends MaskDigitSequence(left, right, "try_mask_ccn", true) {
+  def this(left: Expression) = this(left, Literal(MaskCcn.DEFAULT_FORMAT))
+  override protected def withNewChildrenInternal(
+      newInput: Expression, newFormat: Expression): TryMaskCcn =
+    copy(left = newInput, right = newFormat)
+  override def nullable: Boolean = true
+}
+
+/** Companion object for the MaskCcn and TryMaskCcn classes. */
+object MaskCcn {
+  val DEFAULT_FORMAT = "XXXX-XXXX-XXXX-XXXX"
+}
+
+/** Implementation of an expression to mask digits in a string to replacement characters. */
+abstract class MaskDigitSequence(
+    left: Expression, right: Expression, functionName: String, nullOnError: Boolean)
+  extends BinaryExpression with ImplicitCastInputTypes with NullIntolerant with Serializable {
```

Comment:
why extending Serializable

Reply:
This is needed, it makes the subclass Expression serializable. Removing it causes the tests to fail with serialization exceptions.